src/overmap.cpp

```diff
@@ -75,21 +76,34 @@ oter_iid ot_null,
 std::unordered_map<string_id<oter_t>, oter_t> otermap;
 std::vector<oter_t> oterlist;
 
-struct overmap_location_restictions {
-    std::vector<std::string> allowed;
-    std::vector<std::string> disallowed;
+struct overmap_special_location {
+    std::vector<oter_str_id> allowed;
+    std::vector<oter_str_id> disallowed;
+    // Test if oter meets the terrain restrictions.
+    bool test( const oter_id &oter ) const {
+        const auto matches = [ &oter ]( const oter_str_id &elem ) {
+            return is_ot_type( elem.str(), oter );
+        };
+
+        if( !allowed.empty() && std::none_of( allowed.begin(), allowed.end(), matches ) ) {
+            return false;
+        }
+
+        return std::none_of( disallowed.begin(), disallowed.end(), matches );
+    }
 };
 
 // Map of allowed and disallowed terrain for locations of overmap specials.
-// Format: { location, { { list of allowed terrains }, { list of disallowed terrains } }
+// Format: { location, { { list of allowed terrains }, { list of disallowed terrains } } }
 // @todo Jsonize this map.
-static const std::map<std::string, overmap_location_restictions> overmap_locations = {
-    { "by_road",    { {},                   { "river", "road" } } },
-    { "field",      { { "field" },          {}                  } },
-    { "forest",     { { "forest" },         {}                  } },
-    { "land",       { {},                   { "river", "road" } } },
-    { "water",      { { "river" },          {}                  } },
-    { "wilderness", { { "forest", "field" },{}                  } }
+static const std::map<std::string, overmap_special_location> special_locations = {
+    { "field",      { { oter_str_id( "field"  ) }, {}                         } },
+    { "forest",     { { oter_str_id( "forest" ) }, {}                         } },
+    { "land",       { {},                          { oter_str_id( "river" ),
+                                                     oter_str_id( "road"  ) } } },
```

Comment:
Neither. Swamps are (and were prior to this PR) `"land"` tiles. #19142 adds `"swamp"` as a separate location, and thus we can prohibit `"forest_water"` in `"land"`.